source/server/filter_chain_manager_impl.h

```diff
@@ -71,8 +77,36 @@ class FilterChainFactoryContextImpl : public Configuration::FilterChainFactoryCo
   Configuration::TransportSocketFactoryContext& getTransportSocketFactoryContext() const override;
   Stats::Scope& listenerScope() override;
 
+  void startDraining() { is_draining_.store(true); }
+
 private:
   Configuration::FactoryContext& parent_context_;
+  Init::Manager& init_manager_;
+  std::atomic<bool> is_draining_{false};
+};
+
+class FilterChainImpl : public Network::DrainableFilterChain {
+public:
+  FilterChainImpl(Network::TransportSocketFactoryPtr&& transport_socket_factory,
+                  std::vector<Network::FilterFactoryCb>&& filters_factory)
+      : transport_socket_factory_(std::move(transport_socket_factory)),
+        filters_factory_(std::move(filters_factory)) {}
+
+  // Network::FilterChain
+  const Network::TransportSocketFactory& transportSocketFactory() const override {
+    return *transport_socket_factory_;
+  }
+
+  const std::vector<Network::FilterFactoryCb>& networkFilterFactories() const override {
+    return filters_factory_;
+  }
+
+  void startDraining() override { factory_context_->startDraining(); }
+  std::unique_ptr<FilterChainFactoryContextImpl> factory_context_;
```

Comment:
nit: does this need to be public? Can you make this private and accessed via accessor?